### PR TITLE
Support auxiliary participants in AllReduceProtocol

### LINKED ISSUE #233: Support auxiliary participants in AllReduceProtocol
Currently, DecentralizedAverager supports two types of behavior:
- [regular] peers accumulate gradients AND average them via butterfly all-reduce
- [client_mode] peers accumulate gradients and send them to other peers for averaging

__Problem:__ Colab peers and CS lab desktops require running with client_mode=True due to firewalls. When training with a large number of such peers, one will need a certain amount of non-client-mode participants to aggregate gradients from clients. This is not always affordable, especially when training larger models.

__Proposed solution:__ add a third option for DecentralizedAverager participants: running as an auxiliary node for all-reduce without training the model.

This solution is inspired with the architecture of BytePS system (see Fig.5 in [their paper](https://www.usenix.org/system/files/osdi20-jiang.pdf#page=6)). BytePS authors argue that the presence of such machines can accelerate training even in high-bandwidth setting (Figure 12, same paper). In our case, it will also allow training without cloud GPUs at all.

__Optional implementation tips:__ 
- one way to implement this type of node is to add a special flag into [DecentralizedAverager.step](https://github.com/learning-at-home/hivemind/blob/master/hivemind/client/averaging/__init__.py#L216-L219). E.g. `averager.step(send=False)`. You can probably find a better naming convention.
- This information should then be propagated to AllreduceRunner in _make_allreduce_runner (see `modes` [here](https://github.com/learning-at-home/hivemind/blob/master/hivemind/client/averaging/__init__.py#L276-L291)).
- In turn, AllreduceRunner (and AllreduceProtocol) should support this special role similarly to client_mode_peers.
- If all peers are helpers, protocol should immediately finish without sending tensors around. This is a normal behavior (edge case) that will seldom occur.
- if all peers are client_mode_peers, raise and error

### PR DESCRIPTION
**Pull request by foksly**
Resolve #233

### CODE REVIEW

**Review by mryab — CHANGES_REQUESTED** [review]

File: hivemind/client/averaging/allreduce.py [hunk 1]
```diff
@@ -43,8 +52,14 @@ def __init__(self, *, group_id: GroupID, tensors: Sequence[torch.Tensor], endpoi
         self.averaged_part: asyncio.Future[torch.Tensor] = asyncio.Future()  # will be set to [accumulator / group size]
         self.averaged_tensor_parts: Dict[Endpoint, torch.Tensor] = {}  # averaged chunks from all peers will be put here
         self.future: asyncio.Future[Sequence[torch.Tensor]] = asyncio.Future()  # final result or exception
-        for endpoint in self.client_mode_endpoints:
-            self.averaged_tensor_parts[endpoint] = torch.tensor([])
+        
+        self.num_senders = len([mode for mode in modes if mode != Mode.AUX])
```
[the remark this answers is not in the document]
Comment (foksly) [reply hivemind/client/averaging/allreduce.py:56]: `TypeError: object of type 'generator' has no len()`

File: hivemind/client/averaging/allreduce.py [hunk 2]
```diff
@@ -182,9 +205,13 @@ async def run(self) -> Sequence[torch.Tensor]:
...
         try:
-            await asyncio.gather(self, *(self._communicate_with_peer(peer, self.local_tensor_parts[peer])
-                                         for i, peer in enumerate(self.ordered_group_endpoints)
-                                         if peer not in self.client_mode_endpoints))
+            if self.peer_modes[self.endpoint] != AveragingMode.AUX:
+                await asyncio.gather(self, *(self._communicate_with_peer(peer, self.local_tensor_parts[peer])
+                                            for i, peer in enumerate(self.ordered_group_endpoints)
+                                            if self.peer_modes[peer] != AveragingMode.CLIENT))
+            else:
+                print(f'{self.endpoint} - NOT SENDING STUFF {self.peer_modes}')
```
Comment (mryab) [inline hivemind/client/averaging/allreduce.py:213]: Remove debug print or convert to a logger call (IMO the first is preferable)

File: hivemind/client/averaging/allreduce.py [hunk 3]
```diff
@@ -144,6 +166,7 @@ def _get_peer_stub(self, peer: Endpoint) -> averaging_pb2_grpc.DecentralizedAver
 
     async def _communicate_with_peer(self, peer_endpoint: Endpoint, local_part: torch.Tensor) -> torch.Tensor:
         """ Send a part of local tensors and metadata to a single peer, receive the average for that part of tensors """
+        assert self.peer_modes[self.endpoint] != AveragingMode.AUX, "auxiliary peers are disallowed from sending tensors"
```
Comment (mryab) [inline hivemind/client/averaging/allreduce.py:169]:
```suggestion
        assert self.peer_modes[self.endpoint] != AveragingMode.AUX, "Auxiliary peers are disallowed from sending tensors"
```

File: hivemind/client/averaging/allreduce.py [hunk 4]
```diff
@@ -87,6 +108,7 @@ def register_averaged_part(self, source: Endpoint, averaged_part: torch.Tensor):
         assert source not in self.averaged_tensor_parts, "already registered the average from this peer"
         assert averaged_part.shape == self.local_tensor_parts[source].shape, "averaged part shape mismatch"
         assert averaged_part.dtype == self.local_tensor_parts[source].dtype, "averaged part dtype mismatch"
+        assert self.peer_modes[self.endpoint] != AveragingMode.AUX, "auxiliary peers do not have local tensors for sending"
```
Comment (mryab) [inline hivemind/client/averaging/allreduce.py:111]:
```suggestion
        assert self.peer_modes[self.endpoint] != AveragingMode.AUX, "Auxiliary peers do not have local tensors for sending"
```

File: hivemind/client/averaging/__init__.py [hunk 5]
```diff
@@ -293,19 +301,20 @@ async def _step(self, *, future: MPFuture, gather_binary: bytes, weight: float,
     async def _make_allreduce_runner(self, group_info: GroupInfo, min_vector_size: int, **kwargs) -> AllReduceRunner:
         """ Use a group description found by Matchmaking to form AllreduceRunner """
         try:
-            weights, throughputs, modes, user_gathered = zip(*map(self.serializer.loads, group_info.gathered))
+            weights, throughputs, modes_ix, user_gathered = zip(*map(self.serializer.loads, group_info.gathered))
```
[the remark this answers is not in the document]
Comment (mryab) [reply hivemind/client/averaging/__init__.py:309]: I think it's still worth changing, not sure if `_ix` is a common suffix for indices

**Review by mryab — CHANGES_REQUESTED** [review]

**Review by mryab — APPROVED** [review]

File: hivemind/client/averaging/__init__.py [hunk 6]
```diff
@@ -237,6 +247,8 @@ def step(self, gather: Optional[DataForGather] = None, weight: float = 1.0, time
         :returns: on success, update averaged_tensors and return group info; on failure, return None
         """
         assert isinstance(weight, (int, float)) and weight > 0, f"Expected a positive int/float, got {type(weight)}"
```
Comment (yhn112) [inline hivemind/client/averaging/__init__.py:249]: Seems like this assert should be in the "else" branch of the next "if".

**Review by yhn112 — CHANGES_REQUESTED** [review]

**Comment by foksly** [comment]
Investigation report:
__Problem:__ Test with 2 aux and 2 client peers failed once in a few hundred attempts
__Cause:__ both aux peers entered matchmaking with __exactly__ the same dht_time -- and refused to request each other because of ">=" comparison (to be fixed in subsequent commit)


Example debug prints (from https://github.com/learning-at-home/hivemind/pull/263 )


```
=================================== FAILURES ===================================
___________________________ test_allreduce_once[2-2] ___________________________
n_clients = 2, n_aux = 2

    @pytest.mark.forked
    @pytest.mark.parametrize("n_clients", [0, 1, 2])
    @pytest.mark.parametrize("n_aux", [0, 1, 2])
    def test_allreduce_once(n_clients, n_aux):
>       _test_allreduce_once(n_clients, n_aux)

tests/test_averaging.py:86: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

n_clients = 2, n_aux = 2

    def _test_allreduce_once(n_clients, n_aux):
        dht = hivemind.DHT(start=True, endpoint=f'{hivemind.LOCALHOST}:*')
    
        n_peers = 4
        modes = [AveragingMode.CLIENT] * n_clients + [AveragingMode.AUX] * n_aux + [AveragingMode.NODE] * (n_peers - n_clients - n_aux)
        # random.shuffle(modes)
    
        tensors1 = [torch.randn(123), torch.zeros(3)]
        tensors2 = [torch.rand(123), torch.ones(3)]
        tensors3 = [-torch.rand(123), torch.arange(3).to(torch.float32)]
        tensors4 = [torch.randn(123) ** 3, torch.arange(3).to(torch.float32) / 2]
        peer_tensors = [tensors1, tensors2, tensors3, tensors4]
    
        reference = [sum(tensors[i] for tensors, mode in zip(peer_tensors, modes)
                     if mode != AveragingMode.AUX) / max(1, n_peers - n_aux) for i in range(len(tensors1))]
    
        averagers = [hivemind.DecentralizedAverager(tensors, dht=dht, target_group_size=4, averaging_expiration=15,
                                                    prefix='mygroup', listen=mode != AveragingMode.CLIENT, listen_on='127.0.0.1:*',
                                                    auxiliary=(mode == AveragingMode.AUX), start=True)
                     for tensors, mode in zip(peer_tensors, modes)]
        futures = []
        for averager in averagers:
            futures.append(averager.step(wait=False))
        for future in futures:
            result = future.result()
            for averager in averagers:
>               assert averager.endpoint in result, f"{modes}"
E               AssertionError: [<AveragingMode.CLIENT: 1>, <AveragingMode.CLIENT: 1>, <AveragingMode.AUX: 2>, <AveragingMode.AUX: 2>]
E               assert '127.0.0.1:41813' in {'127.0.0.1:44459': None, 'client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa': None, 'client::59909d38-6f42-4340-adfa-87d57d5f6c3c': None}
E                +  where '127.0.0.1:41813' = DecentralizedAverager(127.0.0.1:41813).endpoint

tests/test_averaging.py:70: AssertionError
----------------------------- Captured stdout call -----------------------------
QQclient::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - current queue = {}
QQclient::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - awaiting queue update
QQclient::59909d38-6f42-4340-adfa-87d57d5f6c3c - current queue = {}
QQclient::59909d38-6f42-4340-adfa-87d57d5f6c3c - awaiting queue update
QQ127.0.0.1:44459 - current queue = {}
QQ127.0.0.1:44459 - awaiting queue updateQQ127.0.0.1:41813 - current queue = {}

QQ127.0.0.1:41813 - awaiting queue update
QQclient::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - current queue = {}
QQclient::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - awaiting queue update
QQ127.0.0.1:44459 - current queue = {}
QQ127.0.0.1:44459 - awaiting queue update
QQclient::59909d38-6f42-4340-adfa-87d57d5f6c3c - current queue = {}
QQclient::59909d38-6f42-4340-adfa-87d57d5f6c3c - awaiting queue update
QQ127.0.0.1:41813 - current queue = {}
QQ127.0.0.1:41813 - awaiting queue update
QQ127.0.0.1:41813 - current queue = {}
QQ127.0.0.1:41813 - awaiting queue update
QQ127.0.0.1:44459 - current queue = {}
QQ127.0.0.1:44459 - awaiting queue update
QQclient::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - current queue = {'127.0.0.1:44459': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQclient::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - yielding 127.0.0.1:44459
client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - REQUEST TO 127.0.0.1:44459
QQclient::59909d38-6f42-4340-adfa-87d57d5f6c3c - current queue = {'127.0.0.1:44459': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQclient::59909d38-6f42-4340-adfa-87d57d5f6c3c - yielding 127.0.0.1:44459
client::59909d38-6f42-4340-adfa-87d57d5f6c3c - REQUEST TO 127.0.0.1:44459
127.0.0.1:44459 - incoming request from client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa (time=1621532332.5730908)
127.0.0.1:44459 - accepted request from client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa
client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - joining the group of 127.0.0.1:44459; waiting for peers
127.0.0.1:44459 - incoming request from client::59909d38-6f42-4340-adfa-87d57d5f6c3c (time=1621532332.5739517)
127.0.0.1:44459 - accepted request from client::59909d38-6f42-4340-adfa-87d57d5f6c3c
client::59909d38-6f42-4340-adfa-87d57d5f6c3c - joining the group of 127.0.0.1:44459; waiting for peers
QQ127.0.0.1:41813 - current queue = {'127.0.0.1:44459': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQ127.0.0.1:41813 - awaiting queue update
QQ127.0.0.1:44459 - current queue = {'127.0.0.1:41813': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQ127.0.0.1:44459 - awaiting queue update
QQ127.0.0.1:41813 - current queue = {'127.0.0.1:44459': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQ127.0.0.1:41813 - awaiting queue update
QQ127.0.0.1:44459 - current queue = {'127.0.0.1:41813': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQ127.0.0.1:44459 - awaiting queue update
QQ127.0.0.1:41813 - current queue = {'127.0.0.1:44459': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
(...repeated several more times...)

QQ127.0.0.1:41813 - awaiting queue update
QQ127.0.0.1:44459 - current queue = {'127.0.0.1:41813': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQ127.0.0.1:44459 - awaiting queue update
QQ127.0.0.1:41813 - current queue = {'127.0.0.1:44459': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQ127.0.0.1:41813 - awaiting queue update
QQ127.0.0.1:44459 - current queue = {'127.0.0.1:41813': ValueWithExpiration(value=1621532347.506267, expiration_time=1621532347.506267)}
QQ127.0.0.1:44459 - awaiting queue update
127.0.0.1:44459 - beginning allreduce because time is up, group: {'client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa': endpoint: "client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa"
schema_hash: "K\250\212\000@B\331\376g\321#k-\336\364*\020<\254~"
expiration: 1621532347.5698905
gather: "\224\313?\360\000\000\000\000\000\000\300\001\304\001\300"
client_mode: true
, 'client::59909d38-6f42-4340-adfa-87d57d5f6c3c': endpoint: "client::59909d38-6f42-4340-adfa-87d57d5f6c3c"
schema_hash: "K\250\212\000@B\331\376g\321#k-\336\364*\020<\254~"
expiration: 1621532347.571211
gather: "\224\313?\360\000\000\000\000\000\000\300\001\304\001\300"
client_mode: true
} (plus 127.0.0.1:44459)target size = 4 (time=1621532347.5079453)
client::59909d38-6f42-4340-adfa-87d57d5f6c3c - beginning alreduce
client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa - beginning alreduce
127.0.0.1:44459 - finished processing request from client::59909d38-6f42-4340-adfa-87d57d5f6c3c
127.0.0.1:44459 - finished processing request from client::2dd4f14d-6b9a-43b2-b754-52b4c29b21fa
QQ127.0.0.1:41813 - current queue = {}
QQ127.0.0.1:41813 - awaiting queue update
QQ127.0.0.1:41813 - current queue = {'127.0.0.1:44459': ValueWithExpiration(value=1621532362.5062153, expiration_time=1621532362.5062153)}
QQ127.0.0.1:41813 - yielding 127.0.0.1:44459
127.0.0.1:41813 - REQUEST TO 127.0.0.1:44459
127.0.0.1:44459 - incoming request from 127.0.0.1:41813 (time=1621532347.5505655)
127.0.0.1:44459 - rejected request from 127.0.0.1:41813
127.0.0.1:41813 - requested 127.0.0.1:44459 to be my leader, but got rejected with NOT_LOOKING_FOR_GROUP
127.0.0.1:44459 - finished processing request from 127.0.0.1:41813
QQ127.0.0.1:41813 - current queue = {}
QQ127.0.0.1:41813 - awaiting queue update
QQ127.0.0.1:41813 - current queue = {}
QQ127.0.0.1:41813 - awaiting queue update

```

**Review by yhn112 — APPROVED** [review]

**Review by justheuristic — APPROVED** [review]

**Pull request merged**